tensorflow/python/kernel_tests/conv_ops_3d_test.py

```diff
@@ -80,23 +90,42 @@ def _SetupValuesForDevice(self, tensor_in_sizes, filter_in_sizes, stride,
   def _VerifyValues(self, tensor_in_sizes, filter_in_sizes, stride, padding,
                     expected):
     results = []
-    for data_format, use_gpu in GetTestConfigs():
-      result = self._SetupValuesForDevice(
-          tensor_in_sizes,
-          filter_in_sizes,
-          stride,
-          padding,
-          data_format,
-          use_gpu=use_gpu)
-      results.append(result)
-      tolerance = 1e-2 if use_gpu else 1e-5
+    for (data_format, use_gpu) in GetTestConfigs():
+      for dtype in self._DtypesToTest(use_gpu):
+        result = self._SetupValuesForDevice(
+            tensor_in_sizes,
+            filter_in_sizes,
+            stride,
+            padding,
+            data_format,
+            dtype,
+            use_gpu=use_gpu)
+        results.append((result, use_gpu, dtype, data_format))
+
       with self.test_session() as sess:
-        values = sess.run(results)
-        for value in values:
+        tensor_results = [result[0] for result in results]
+        test_params = [result[1:] for result in results]
+        values = sess.run(tensor_results)
+        for i in range(len(tensor_results)):
+          value = values[i]
+          print("use_gpu:", test_params[i][0])
+          print("dtype:", test_params[i][1])
+          print("data_format:", test_params[i][2])
+
           print("expected = ", expected)
           print("actual = ", value)
-          self.assertAllClose(expected, value.flatten(), atol=tolerance,
-                              rtol=1e-6)
+          tol = 1e-5
+          if value.dtype == np.float16:
+            tol = 1e-3
+            # fp16 using may result in inf values and large absolute errors
+            # when used with large numbers
```

Comment:
Could you be more specific about large numbers? Or if you could change the value of the input tensors to make them smaller, that would be better. I'm worried that the skipped tests may affect our coverage.

Reply:
fp16 max value is `65504.0`. So if we try to convolve array and filter which values contain 1e3 it's highly probable that I will got `inf`. Should I write this in code comment?

For now I use test values from old tests for both dtypes and skip only cases with large values. I can try to divide large numbers in fp16 tests. Is it appropriate?

Reply:
Sorry, I mean "which contain values greater than 1e4", as checked here
 https://github.com/opensourcemattress/tensorflow/blob/0853d075ceaaaf18f5d8de35656635f9e4cd6845/tensorflow/python/kernel_tests/conv_ops_3d_test.py#L122

Reply:
Your explanation here is fine. Please remove the additional logging lines (3 prints) in there (111,112,113). Thank you!

Reply:
I don't think we have enough tests due to the fact we skip large numbers. For example, testConv3D2x2x2FilterStride2, testConv3DStride3, and testConv3D2x2x2FilterStride2Same are not run at all with fp16. Can you add a few more tests, testing the sorts of things that are currently skipped, that are run with fp16?